flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalogBase.java

```diff
@@ -89,6 +101,34 @@ private static IMetaStoreClient getMetastoreClient(HiveConf hiveConf) {
 		}
 	}
 
+	// ------ APIs ------
+
+	/**
+	 * Validate input base table.
+	 *
+	 * @param catalogBaseTable the base table to be validated
+	 * @throws IllegalArgumentException thrown if the input base table is invalid.
+	 */
+	protected abstract void validateCatalogBaseTable(CatalogBaseTable catalogBaseTable)
```

Comment:
Even though this is not a public API, I feel it should be more formal. Specifically, we should probably can let this return true or false for validation result, rather than relying on throwing IllegalArgumentException. Secondly, validation can be more more complicated than just check the instance type. At this level, it shouldn't care what kind of validation the sub classes might do.

I think it's acceptable if the base class doesn't care if the sub class does validation or how it does it. So it's okay if we remove this and let each subclass do its validation when create/alter table. This way might be even cleaner.

Reply:
This would go against the previous suggestion you made:

> Similar to createTable(), I don't see we need to provide implementations for alterTable() in each subclass because I don't really see much difference. If validation is different, then we can define a new interface in the base and have each subclass implement it.

Reply:
I'm not backtracking what I said earlier. It's fine to defined an API, but the API needs to be well-thought, which doesn't seem to be the case here. That's why I said we can forgo the API. If we think an API is the way to go, then let's define the API better and parent class shouldn't define what validation a subclass might do. Personally, I don't feel the API is sufficiently sound, but feel free to get 2nd opinion though.

Reply:
Make sense. Though IllegalArgumentException can be thrown when any part of the input table is invalid, a better signature may just be throwing CatalogException to conform with that of other existing catalog APIs, since we've defined CatalogException's use case as "in case of any runtime exception".

I don't think returning boolean is a good signature as the check can be fairly complicated and the caller won't know which part is invalid. Returning boolean would be a good option for a single criteria check like "bool isTable()".

Reply:
If you're concerned about giving user precise, useful information, throwing an exception isn't the best choice. For that, logging is probably better. Neverthelesss, CatalogException certainly sounds better, though it doesn't change the nature of the problem. Let's do this way for now.